sys/kern/vm_map.c

```diff
@@ -425,39 +439,75 @@ void vm_map_dump(vm_map_t *map) {
          (it->prot & VM_PROT_READ) ? 'r' : '-',
          (it->prot & VM_PROT_WRITE) ? 'w' : '-',
          (it->prot & VM_PROT_EXEC) ? 'x' : '-');
-#if 0
-    vm_object_dump(it->object);
-#endif
   }
 }
 
+static vm_map_entry_t *vm_map_entry_clone_shared(vm_map_t *map,
+                                                 vm_map_entry_t *ent) {
+  if (!ent->aref.amap) {
+    /* We need to create amap, because we won't be able to share it if it
+     * is not created now. */
+    int slots = vaddr_to_slot(ent->end - ent->start);
+    ent->aref.amap = vm_amap_alloc(slots);
+    if (!ent->aref.amap)
+      return NULL;
+  }
+
+  vm_map_entry_t *new = vm_map_entry_copy(ent);
+  vm_amap_hold(ent->aref.amap);
+  new->aref = ent->aref;
+  return new;
+}
+
+static vm_map_entry_t *vm_map_entry_clone_copy(vm_map_t *map,
+                                               vm_map_entry_t *ent) {
+  vm_map_entry_t *new = vm_map_entry_copy(ent);
+
+  /* If there was an amap we need to clone it. */
+  if (ent->aref.amap) {
```

Comment:
Please use early exit in here as the if-clause constitutes majority of the work performed by the function.